### PR TITLE
[RNMobile] Stepper component

### PR DESCRIPTION
**Pull request by geriux**
`Gutenberg-mobile` PR -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/1640

## Description
This PR adds a new mobile component `Stepper`

**Note**: To be able to test this, I've replaced the `SliderControl` with this new component `StepperControl`. I'll revert this before merging.

### To test

- Open the app
- Add the `Spacer` block
- Open its settings
- Change the `Height` using the stepper control
- **Expect**: value should change by pressing in or just with a normal tap.
- Accessibility test:
  - **iOS**:
    - Turn on VoiceOver
    - Select the Stepper control
    - **Expect**: value should change by swiping up / down with one finger. Sound with the value change should play after every change.
  - **Android**:
    - Turn on TalkBack
    - With the settings open: Swipe with one finger from left to right to navigate throughout the UI elements and set focus on the stepper control
    - Once it's focused, use volume keys to change the value
    - **Expect**: value changing and sound with the change being played.

| iOS | Android |
|     :---:      |     :---:      |
|  |   |

| Accessibility |
|     :---:      | 
|    |

For **Android** devices, users will be able to change the value using the Up / Down volume buttons. It will announce the new value with any change.

For **iOS**, users will be able to swipe up / down with one finger on the screen to change the value. It will also announce the new value after a change.

### iOS 

| Default | Dark mode |
|     :---:      |     :---:      |
|  |   |

 | iPad |
|     :---:      | 
|    |

---

### Android

 

---

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [x] I've included developer documentation if appropriate.
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/spacer/edit.native.js
Comment (geriux) [inline packages/block-library/src/spacer/edit.native.js]: This is for testing only.  I will rever this before merging.

File: packages/components/src/mobile/bottom-sheet/stepper-cell/README.md
Comment (etoledom) [inline packages/components/src/mobile/bottom-sheet/stepper-cell/README.md]: This kind of docs looks great! Thank you for including it.

But this is something I'm not too familiar with, for example:
`separatorType` is a prop of the base `Cell` component. As far as I understand, in this file we should document the props that are particular for the `stepper` component, and `inherit` properties (even from docs) from the parent component.

Does it makes sense or maybe I'm lost? :)

File: packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js
Comment (etoledom) [inline packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js]: Great solution for accessibility, works like a charm ✨

File: packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js
Comment (etoledom) [inline packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js]: Not sure if we should use `_x` here.

As far as I understand, `_x` is used when the same frase might need different translations depending on the context. We do have another instance of this same string, but not sure if the context is different (this is a picker while the other is the slider).

https://github.com/WordPress/gutenberg/blob/2ebfb64ce589f743b7caa8a8e1727b576755e16e/packages/components/src/mobile/bottom-sheet/range-cell.native.js#L152

If we keep it, I'd suggest to use a similar context string, something like: `Picker for selecting a number inside a range`.

File: packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js
Comment (etoledom) [inline packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js]: From what I see [here](https://facebook.github.io/react-native/docs/accessibility#accessibility-actions), labels for standard actions (like increment/decrement) are not required, but if we add them, the best would be to localise them. I'd suggest to not add them in this case and let the system do their magic.

File: docs/manifest-devhub.json
Comment (etoledom) [inline docs/manifest-devhub.json]: Not sure if we should add this here 🤔 
`StepperCell` is an internal component that we make available via the native version of `StepperControl`.

@gziolo any light on this?

**Review by etoledom — COMMENTED** [review]
Great job @geriux ! It works and look great ✨ 
I really like the accessibility solution, feels so natural 👍 

I added a few small code comments, but nothing really important.

There's also one extra detail that I think we could improve. When the cell is pressed, it reacts with opacity. Other cells have some logic to this but this one does nothing:

cell

What do you think about passing a `disabled=true` property down to the cell's base `TouchableOpacity` and avoid this opacity animation from happening?

**Comment by etoledom** [comment]
One more thing! 😁

Tested a bigger text size on Android and the layout broke:
Screenshot_20191210-100104

I wouldn't worry on this iteration about the huge iOS Accessibility font sizes (since it probably will be quite troublesome to fix) but this font sizes (just a bit bigger) are quite common. It would be great to check this out.

File: docs/manifest-devhub.json
Comment (gziolo) [reply docs/manifest-devhub.json]: It's interesting that it gets detected. I guess this is how it works with `README.md` files added to `@wordpress/components`.

I don't think it's internal as it gets exposed and consumed in the block library. We should just make it clear that it will work only with the React Native code. @youknowriad - any thoughts on how to make it possible to ignore all mobile docs from being exposed in the documentation published at https://developers.wordpress.org?

File: packages/components/src/mobile/bottom-sheet/stepper-cell/README.md
Comment (pinarol) [reply packages/components/src/mobile/bottom-sheet/stepper-cell/README.md]: I think we should add the docs for the StepperControl instead of StepperCell, because StepperControl is the public API. And for inheriting the docs, it can only be possible if parent view is also part of the public `@wordpress/components` API. Base `Cell` is not public and i think it won't be public. If we had long term plans for `separatorType` I'd suggest adding a public parent view but we want to [get rid of that](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1421) one as soon as possible so not sure if it's worth it.

Comment (etoledom) [reply packages/components/src/mobile/bottom-sheet/stepper-cell/README.md]: > I think we should add the docs for the StepperControl instead of StepperCell

That's what I thought at the beginning, but I was surprised that there wasn't docs already for `StepperCell`.

Also `icon` and `label` are also inherited props. Maybe we want to use BaseControl as an abstraction of our base Cell component (and also extract the TextInput logic from it).

File: docs/manifest-devhub.json
Comment (etoledom) [reply docs/manifest-devhub.json]: > I don't think it's internal as it gets exposed and consumed in the block library.

We are currently using the controls abstractions (StepperControl, RangeControl, etc...), so the mobile-only `-Cell` components are not being consumed by any other libraries.

We are planning on adding these docs directly to StepperControl instead. Wdyt?

File: packages/components/src/mobile/bottom-sheet/stepper-cell/README.md
Comment (geriux) [reply packages/components/src/mobile/bottom-sheet/stepper-cell/README.md]: Thanks for the feedback! It makes sense to move the docs to `StepperControl` and remove the props that are not specific for the component itself. I'll start with that but it's also good to know how we could improve it in the future.

File: packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js
Comment (geriux) [reply packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js]: Oh right! we do have the Slider one. I'll change the `_x` thanks!!

File: packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js
Comment (geriux) [reply packages/components/src/mobile/bottom-sheet/stepper-cell/index.native.js]: Sounds good! Way better to avoid handling those localizations =D

**Comment by geriux** [comment]
> Great job @geriux ! It works and look great ✨
> I really like the accessibility solution, feels so natural 👍

Thank you! I think so too! Tested different methods and this one seemed just right 😁

> What do you think about passing a `disabled=true` property down to the cell's base `TouchableOpacity` and avoid this opacity animation from happening?

Good idea! I'll check it out 😃

**Comment by geriux** [comment]
@etoledom Updated with some changes =)

**Review by etoledom — APPROVED** [review]
Thank you for the update @geriux - It's working great now 🎉 

Let's wait for @gziolo and/or @youknowriad to ✅ the documentation bit.
The rest looks good to me. Great job!

**Comment by geriux** [comment]
> Thank you for the update @geriux - It's working great now 🎉
> 
> Let's wait for @gziolo and/or @youknowriad to ✅ the documentation bit.
> The rest looks good to me. Great job!

Yay! Sounds good! Thank you! 🎉

**Comment by gziolo** [comment]
I have two very basic questions to better understand the purpose of this new component.

What's the difference between this component and `RangeControl`?

See:
- https://github.com/WordPress/gutenberg/tree/master/packages/components/src/range-control
- https://wordpress.github.io/gutenberg/?path=/story/components-rangecontrol--default

Why is it specific to mobile only?

**Comment by pinarol** [comment]
@gziolo Basically we want the Stepper component to represent controls that has ≤ 6 or 8 options instead of Slider. It is a UX request made by @iamthomasbishop so I'll defer to him to provide better insight about this.

**Comment by iamthomasbishop** [comment]
> It is a UX request made by @iamthomasbishop so I'll defer to him to provide better insight about this.

Yes, as @pinarol mentioned Stepper (or even Picker, which @koke has mentioned) is better suited than a Slider for controls with a small range of options. For example, Apple's [HIG](https://developer.apple.com/design/human-interface-guidelines/ios/controls/steppers/) suggests the following:

> Steppers work well for making small changes that require a few taps.

**Comment by gziolo** [comment]
I opened https://github.com/WordPress/gutenberg/pull/19236 to ensure that mobile-only components aren't exposed in WordPress developers' docs. It should make it easier for you to keep all the work documented.

**Pull request merged**
